pkg/controller/virtualmachine/virtualmachine_controller.go

```diff
@@ -86,15 +85,55 @@ func (r *ReconcilePolicy) Reconcile(request reconcile.Request) (reconcile.Result
 	err := r.Get(context.TODO(), request.NamespacedName, instance)
 	if err != nil {
 		if errors.IsNotFound(err) {
-			err := r.poolManager.ReleaseVirtualMachineMac(fmt.Sprintf("%s/%s", request.Namespace, request.Name))
-			if err != nil {
-				logging.Printf(logging.ErrorLevel, "failed to release mac for virtual machine %s: %v", request.NamespacedName, err)
-			}
+
 			return reconcile.Result{}, nil
 		}
 		// Error reading the object - requeue the request.
 		return reconcile.Result{}, err
 	}
+	if !instance.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
we also need to section here to add the finalizers because when we start the cluster maybe already contains pod/vms and we need to add the finalizer to them.